File: app/src/main/java/com/kite/mnemoai/fragment/dialog/VocabularySelectDialogFragment.java
package com.kite.mnemoai.fragment.dialog;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.kite.mnemoai.R;
import com.kite.mnemoai.adapter.VocabularySelectAdapter;
import com.kite.mnemoai.databinding.DialogFragmentVocabularySelectBinding;

import java.util.List;
import java.util.stream.Collectors;

public class VocabularySelectDialogFragment extends DialogFragment {
    private List<VocabularySelectInfo> vocabularySelectInfos;
    private IVocabularySelectListener vocabularySelectListener;

    public VocabularySelectDialogFragment(List<VocabularySelectInfo> vocabularySelectInfos, IVocabularySelectListener vocabularySelectListener){
        this.vocabularySelectInfos = vocabularySelectInfos;
        this.vocabularySelectListener = vocabularySelectListener;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        assert getParentFragment() != null;
        AlertDialog.Builder builder = new AlertDialog.Builder(getParentFragment().requireContext());
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        DialogFragmentVocabularySelectBinding binding = DialogFragmentVocabularySelectBinding.inflate(inflater, null, false);
        RecyclerView vocabularyRV = binding.vocabularyRV;
        VocabularySelectAdapter vocabularySelectAdapter = new VocabularySelectAdapter(vocabularySelectInfos);
        vocabularyRV.setLayoutManager(new LinearLayoutManager(getContext()));
        vocabularyRV.setAdapter(vocabularySelectAdapter);
        builder.setView(binding.getRoot())
                .setPositiveButton(R.string.confirm, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        List<Long> ids = vocabularySelectInfos.stream()
                                .filter(vocabularySelectInfo -> vocabularySelectInfo.isSelect)
                                .map(VocabularySelectDialogFragment.VocabularySelectInfo::getId)
                                .collect(Collectors.toList());
                        vocabularySelectListener.onConfirm(ids);
                    }
                })
                .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        vocabularySelectListener.onNavigate(VocabularySelectDialogFragment.this);
                    }
                });
        return builder.create();
    }

    public List<VocabularySelectInfo> getVocabularySelectInfos() {
        return vocabularySelectInfos;
    }

    public interface IVocabularySelectListener{
        public void onConfirm(List<Long> ids);
        public void onNavigate(DialogFragment dialogFragment);
    }

    public static class VocabularySelectInfo{
        private long id;
        private String name;
        private String describe;
        private boolean isSelect;

        public VocabularySelectInfo(long id, String name) {
            this.id = id;
            this.name = name;
        }

        public String getDescribe() {
            return describe;
        }

        public void setDescribe(String describe) {
            this.describe = describe;
        }

        public long getId() {
            return id;
        }

        public void setId(long id) {
            this.id = id;
        }

        public boolean isSelect() {
            return isSelect;
        }

        public void setSelect(boolean select) {
            isSelect = select;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }
    }
}

File: app/src/main/java/com/kite/mnemoai/ui/dialog/vocabularyselect/VocabularySelectDialogFragment.kt
package com.kite.mnemoai.ui.dialog.vocabularyselect

import android.app.Dialog
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.kite.mnemoai.R
import com.kite.mnemoai.adapter.VocabularySelectAdapter
import com.kite.mnemoai.databinding.DialogFragmentVocabularySelectBinding

class VocabularySelectDialogFragment: DialogFragment() {
    private lateinit var vocabularySelectInfos: MutableList<VocabularySelectInfo?>
    companion object{
        const val VOCABULARY_BOOK_SELECT = "VOCABULARYBOOKSELECT"

        fun newInstance(vocabularySelectInfos: MutableList<VocabularySelectInfo?>): VocabularySelectDialogFragment{
            val bundle = Bundle().apply {
                putParcelableArrayList("vocabulary_select_infos", ArrayList(vocabularySelectInfos))
            }
            val dialogFragment = VocabularySelectDialogFragment()
            dialogFragment.arguments = bundle
            return dialogFragment
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        vocabularySelectInfos =
            arguments?.getParcelableArrayList<VocabularySelectInfo>(
                "vocabulary_select_infos",
            ) ?: mutableListOf()
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return activity?.let {
            val binding = DialogFragmentVocabularySelectBinding.inflate(layoutInflater, null, false)
            val vocabularySelectAdapter = VocabularySelectAdapter(vocabularySelectInfos)
            binding.vocabularyRV.apply {
                setLayoutManager(LinearLayoutManager(context))
                setAdapter(vocabularySelectAdapter)
            }
            binding.cancelBtn.setOnClickListener {
                dismiss()
            }
            binding.confirmBtn.setOnClickListener {
                val ids = vocabularySelectInfos
                    .filter { vocabularySelectInfo -> vocabularySelectInfo?.isSelect == true }
                    .mapNotNull { vocabularySelectInfo -> vocabularySelectInfo?.id }
                    .toLongArray()

                val bundle = Bundle().apply {
                    putLongArray("ids", ids)
                }
                parentFragmentManager.setFragmentResult(VOCABULARY_BOOK_SELECT, bundle)
                dismiss()
            }
            val builder = MaterialAlertDialogBuilder(it, R.style.CustomAlertDialog)
            builder.setView(binding.root)
            builder.create()
        }?:throw IllegalStateException("页面不存在")
    }

    data class VocabularySelectInfo(var id: Long, @JvmField var name: String?, @JvmField var description: String?, ):
        Parcelable {

        constructor(parcel: Parcel) : this(
            parcel.readLong(),
            parcel.readString() ?: "",
            parcel.readString() ?: ""
        )
        @JvmField
        var isSelect: Boolean = false
        override fun describeContents(): Int {
            return 0
        }

        override fun writeToParcel(p0: Parcel, p1: Int) {
            p0.apply {
                writeLong(id)
                writeString(name)
                writeString(description)
                writeInt(if (isSelect) 1 else 0)
            }
        }

        companion object CREATOR:  Parcelable.Creator<VocabularySelectInfo>{
            override fun createFromParcel(p0: Parcel?): VocabularySelectInfo? {
                return p0?.let { VocabularySelectInfo(p0) }
            }

            override fun newArray(p0: Int): Array<out VocabularySelectInfo?>? {
                return arrayOfNulls(p0)
            }

        }
    }
}
